flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,296 @@
+		this.pubSubSubscriberFactory = pubSubSubscriberFactory;
+		this.credentials = credentials;
+		this.projectSubscriptionName = projectSubscriptionName;
+		this.maxMessagesPerPull = maxMessagesPerPull;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		this.eventLoopGroup = new NioEventLoopGroup();
+		this.subscriber = pubSubSubscriberFactory.getSubscriber(eventLoopGroup, credentials);
+		this.deduplicateMessages = getRuntimeContext().getNumberOfParallelSubtasks() == 1;
+		this.isRunning = true;
+		this.pullRequest = PullRequest.newBuilder()
+										.setMaxMessages(100)
+										.setReturnImmediately(false)
+										.setSubscription(projectSubscriptionName)
+										.build();
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<String> acknowledgementIds) {
+		if (acknowledgementIds.isEmpty() || !isRunning) {
+			return;
+		}
+
+		AcknowledgeRequest acknowledgeRequest =
+			AcknowledgeRequest.newBuilder()
+								.setSubscription(projectSubscriptionName)
+								.addAllAckIds(acknowledgementIds)
+								.build();
+		subscriber.acknowledgeCallable().call(acknowledgeRequest);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		while (isRunning) {
+			messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
+			try {
+				processMessage(sourceContext, messagesFuture.get().getReceivedMessagesList());
+			} catch (InterruptedException | CancellationException e) {
+				awaitSubscriberTermination();
+				return;
+			}
+		}
+
+		awaitSubscriberTermination();
+	}
+
+	void processMessage(SourceContext<OUT> sourceContext, List<ReceivedMessage> messages) throws IOException {
+		synchronized (sourceContext.getCheckpointLock()) {
+			for (ReceivedMessage message : messages) {
+				sessionIds.add(message.getAckId());
+
+				PubsubMessage pubsubMessage = message.getMessage();
+				if (deduplicateMessages && !addId(pubsubMessage.getMessageId())) {
+					// message is duplicate so just ignore it
+					return;
+				}
+
+				OUT deserializedMessage = deserializeMessage(pubsubMessage);
+				if (deserializationSchema.isEndOfStream(deserializedMessage)) {
+					stop();
+					return;
+				}
+
+				sourceContext.collect(deserializedMessage);
+			}
+
+		}
+	}
+
+	private Integer getOutstandingMessagesToAck() {
```

Comment:
The Gauge value will be read in a different thread. So `sessinIdsPerSnapshot` needs to be thread safe. 

Also, usually we should avoid computation in the metric reading. This is because the background metric reading thread may have many metrics to deal with, computation in reading may slowdown that thread and result in metric problems.

Reply:
There's not a lot of computation being done in this function though, All it does is sum a list per uncompleted checkpoint.

I don't know how often uncompleted checkpoints are present generally, but in my own jobs there's never more than 1. 

Do you think this will be a problem?

Reply:
Maybe the amount of computation is not a big concern here. But since there are two threads accessing `sessionIdsPerSnapshot`, the thread safety issue is still a concern.

Reply:
The computation might not be a big concern here. But should the thread safety issue be addressed to avoid `ConcurrentModificationException` when two threads access `sessionIdsPerSnapshot` at the same time?

Reply:
Yeah you're right, it seems to me that the problem exists in the notifyCheckpointComplete() as well. In that function it will remove elements from `sessionIdsPerSnapshot`. 

I will change it from `ArrayDeque` to a `ConcurrentLinkedDeque`. The `List` inside of `sessionIdsPerSnaphot` is fine because we only call `size()` which works on a int.

This should fix it right?

Reply:
Yes, that should work.

Reply:
It seems the `ArrayDeque` is still used instead of the `ConcurrentLinkedDeque`?